File: Spark_Recommendation_content/task1.scala
import org.apache.spark.sql.SparkSession
import scala.util.Random
import scala.math.sqrt
import java.io._


object Task1 {

  def isPrime(n: Int): Boolean = {
    if (n <= 1) false
    else (2 to math.sqrt(n).toInt).forall(n % _ != 0)
  }

  def nextPrime(n: Int): Int = {
    var candidate = n
    while (!isPrime(candidate)) {
      candidate += 1
    }
    candidate
  }

  def computeSignature(userSet: Set[String],
                       userIndex: Map[String, Long],
                       hashFuncs: Seq[Long => Long]): Array[Long] = {

    val sig = Array.fill[Long](hashFuncs.length)(Long.MaxValue)
    for (user <- userSet) {
      userIndex.get(user).foreach { idx =>
        for ((f, i) <- hashFuncs.zipWithIndex) {
          val hVal = f(idx)
          if (hVal < sig(i)) sig(i) = hVal
        }
      }
    }
    sig
  }

  def splitIntoBands(businessId: String, signature: Array[Long], b: Int, r: Int): Seq[((Int, Int), String)] = {
    (0 until b).map { i =>
      val band = signature.slice(i * r, (i + 1) * r).toSeq

      val bandKey = (i, band.hashCode())
      (bandKey, businessId)
    }
  }

  def generateCandidatePairs(businessList: Seq[String]): Seq[(String, String)] = {
    val sortedList = businessList.sorted
    for {
      i <- sortedList.indices
      j <- (i + 1) until sortedList.size
    } yield (sortedList(i), sortedList(j))
  }

  def jaccardSimilarity(users1: Set[String], users2: Set[String]): Double = {
    val intersectionSize = (users1 intersect users2).size
    val unionSize = (users1 union users2).size
    if (unionSize == 0) 0.0 else intersectionSize.toDouble / unionSize
  }

  def main(args: Array[String]): Unit = {
    if (args.length < 2) {
      println("Usage: Task1 <input_file> <output_file>")
      sys.exit(1)
    }
    val inputFile = args(0)
    val outputFile = args(1)

    val spark = SparkSession.builder().appName("Task1").getOrCreate()
    val sc = spark.sparkContext

    val rdd = sc.textFile(inputFile)
    val header = rdd.first()
    
    val dataRdd = rdd.filter(line => line != header)

    val kvRdd = dataRdd.map(line => {
      val fields = line.split(",")
      (fields(1), fields(0))
    })

    val businessSets = kvRdd.mapValues(user => Set(user)).reduceByKey(_ union _)
    val allUsers = kvRdd.map(_._2).distinct()
    val userPermutation = allUsers.zipWithIndex().collectAsMap()


    val numUsers = userPermutation.size
    val m = nextPrime(numUsers)
    val numHashFuncs = 100
    val rand = new Random()
    //  f(x) = (a*x + b) % m 
    val hashFuncs: Seq[Long => Long] = (0 until numHashFuncs).map { _ =>
      val a = rand.nextInt(m - 1) + 1
      val b = rand.nextInt(m)
      (x: Long) => (a * x + b) % m
    }

    // (business_id, signature: Array[Long])
    val businessSignatures = businessSets.mapValues(users => computeSignature(users, userPermutation.toMap, hashFuncs))

    //
    val bNum = 20
    val rNum = 5
    val bandsRdd = businessSignatures.flatMap { case (businessId, signature) =>
      splitIntoBands(businessId, signature, bNum, rNum)
    }
    
    val groupedBands = bandsRdd.groupByKey().mapValues(_.toSeq)
    // 生成候选对：对每个 band 的业务列表，两两组合
    val candidatePairs = groupedBands.flatMap { case (_, businessList) =>
      generateCandidatePairs(businessList)
    }.distinct()

    val businessUsersDict = sc.broadcast(businessSets.collectAsMap())


    val candidateSimilarities = candidatePairs.map { case (b1, b2) =>
      val users1 = businessUsersDict.value(b1)
      val users2 = businessUsersDict.value(b2)
      val sim = jaccardSimilarity(users1, users2)
      (b1, b2, sim)
    }

    val similarPairsSorted = candidateSimilarities.filter(_._3 >= 0.5)
      .sortBy { case (b1, b2, sim) => (b1, b2, sim) }

    val sortedList = similarPairsSorted.collect().toList

  
  
    val writer = new PrintWriter(new File(outputFile))
    writer.println("business_id_1,business_id_2,similarity")
    sortedList.foreach { case (b1, b2, sim) =>
      writer.println(s"$b1,$b2,$sim")
    }
    writer.close()

    spark.stop()
  }
}

File: Spark_Recommendation_content/task1.py
from pyspark.sql import SparkSession
import json
import sys
import random



def is_prime(n):
    if n <= 1:
        return False
    for i in range(2, int(n**0.5) + 1):
        if n % i == 0:
            return False
    return True

def next_prime(n):
    while True:
        if is_prime(n):
            return n
        n += 1



def compute_signature(user_set, user_index_dict, hash_funcs):
    
    signature = [float('inf')] * len(hash_funcs)
    for user in user_set:
        if user in user_index_dict:
            user_index = user_index_dict[user]
            for i, f in enumerate(hash_funcs):
                h_val = f(user_index)
                if h_val < signature[i]:
                    signature[i] = h_val
    return signature

def split_into_bands(business_id, signature, b, r):
    bands = []
    for i in range(b):
        band = tuple(signature[i * r : (i + 1) * r])
        band_key = (i, hash(band))
        bands.append((band_key, business_id))
    return bands

def generate_candidate_pairs(business_list):
    business_list = sorted(business_list) 
    pairs = []
    n = len(business_list)
    for i in range(n):
        for j in range(i+1, n):
            pairs.append((business_list[i], business_list[j]))
    return pairs


def main(input_file, output_filepath):
    spark = SparkSession.builder.appName("task1").getOrCreate()
    sc = spark.sparkContext

    rdd  = sc.textFile(input_file)

    header = rdd.first()

    kv_rdd = rdd.filter(lambda x: x != header).map(lambda x: x.split(',')).map(lambda x: (x[1], x[0]))


    # convert to set in order to speed up the i/o 
    business_sets = kv_rdd.mapValues(lambda x: {x}).reduceByKey(lambda set1, set2: set1 | set2)

    # give a fix permutation
    all_users = kv_rdd.map(lambda x: x[1]).distinct()

    user_permutation = all_users.zipWithIndex().collectAsMap()

    # setup hash function
    num_users = len(user_permutation)
    m = next_prime(num_users)
    num_hash_funcs = 100 

    hash_funcs = []
    for _ in range(num_hash_funcs):
        a = random.randint(1, m - 1)
        b = random.randint(0, m - 1)

        hash_funcs.append(lambda x, a=a, b=b: (a * x + b) % m)

    business_signatures = business_sets.mapValues(
        lambda users: compute_signature(users, user_permutation, hash_funcs)
    )

    b = 20
    r = 100 // b

    bands_rdd = business_signatures.flatMap(lambda x: split_into_bands(x[0], x[1], b, r))

    grouped_bands = bands_rdd.groupByKey().mapValues(list)


    candidate_pairs = grouped_bands.flatMap(lambda x: generate_candidate_pairs(x[1])).distinct()


    business_users_dict = sc.broadcast(business_sets.collectAsMap())

    def jaccard_similarity(pair):
        b1, b2 = pair
        users1 = business_users_dict.value[b1]
        users2 = business_users_dict.value[b2]
        intersection = len(users1.intersection(users2))
        union = len(users1.union(users2))
        return float(intersection) / union if union != 0 else 0.0


    candidate_similarities = candidate_pairs.map(lambda pair: (pair[0], pair[1], jaccard_similarity(pair)))


    similar_pairs_sorted = candidate_similarities.filter(lambda x: x[2] >= 0.5).sortBy(lambda x: (x[0], x[1], x[2]))


    sorted_list = similar_pairs_sorted.collect()


    with open(output_filepath, "w", encoding="utf-8") as f:
        f.write("business_id_1,business_id_2,similarity\n")
        for (b1, b2, sim) in sorted_list:
            f.write(f"{b1},{b2},{sim}\n")

    spark.stop()



if __name__ == "__main__":

    input_file = sys.argv[1]
    out_file = sys.argv[2]
    main(input_file, out_file)


